web/src/lib/utils/cast-player.ts

```diff
@@ -0,0 +1,276 @@
+        this.remotePlayer.set(event.value);
+        break;
+      }
+      case 'duration': {
+        this.duration.set(event.value);
+        break;
+      }
+      case 'currentTime': {
+        this.currentTime.set(event.value);
+        break;
+      }
+      case 'playerState': {
+        this.playerState.set(event.value);
+        break;
+      }
+    }
+  }
+
+  async loadMedia(mediaUrl: string) {
+    if (!get(this.isInitialized)) {
+      return;
+    } else if (!this.session) {
+      return;
+    } else if (this.currentUrl === mediaUrl) {
+      // Media already loaded, prevent reloading
+      return;
+    } else if (get(this.castState) !== cast.framework.CastState.CONNECTED) {
+      // Cast not connected
+      return;
+    }
+
+    if (!this.apiKey) {
+      const apiKey = await this.getCastApiKey();
+
+      if (!apiKey) {
+        console.error('No cast api available');
+        return;
+      }
+
+      this.apiKey = apiKey;
+    }
+
+    let contentType: string | null = null;
+
+    await fetch(mediaUrl, { method: 'HEAD' }).then((response) => {
+      contentType = response.headers.get('content-type');
+    });
+
+    if (!contentType) {
+      console.error('Could not get content type for url ' + mediaUrl);
+      return;
+    }
+
+    const authenticatedUrl = `${mediaUrl}&apiKey=${this.apiKey.secret}`;
+    const mediaInfo = new chrome.cast.media.MediaInfo(authenticatedUrl, contentType);
+
+    const request = new chrome.cast.media.LoadRequest(mediaInfo);
+
+    const successCallback = this.onMediaDiscovered.bind(CastPlayer.instance, SESSION_DISCOVERY_CAUSE.LOAD_MEDIA);
+    this.currentUrl = mediaUrl;
+    return this.session.loadMedia(request, successCallback, this.onError.bind(this));
+  }
+
+  onError(error: chrome.cast.Error) {
+    console.error('cast player error', error);
+  }
+
+  onMediaDiscovered(cause: SESSION_DISCOVERY_CAUSE, currentMedia: chrome.cast.media.Media) {
+    this.currentMedia = currentMedia;
+
+    if (cause === SESSION_DISCOVERY_CAUSE.LOAD_MEDIA) {
+      this.castPlayerState.set(chrome.cast.media.PlayerState.PLAYING);
+    } else if (cause === SESSION_DISCOVERY_CAUSE.ACTIVE_SESSION) {
+      this.castPlayerState.set(currentMedia.playerState);
+    }
+
+    this.currentMedia.addUpdateListener(this.onMediaStatusUpdate.bind(this));
+  }
+
+  onMediaStatusUpdate(mediaStillAlive: boolean) {
+    if (mediaStillAlive) {
+      const friendlyName = this.session?.receiver.friendlyName;
+      if (friendlyName) {
+        this.receiverFriendlyName.set(friendlyName);
+      } else {
+        this.receiverFriendlyName.set(null);
+      }
+    } else {
+      this.castPlayerState.set(chrome.cast.media.PlayerState.IDLE);
+    }
+  }
+
+  sessionMediaListener(currentMedia: chrome.cast.media.Media) {
+    this.currentMedia = currentMedia;
+    this.currentMedia.addUpdateListener(this.onMediaStatusUpdate.bind(this));
+  }
+
+  private async createCastApiKey() {
+    try {
+      const data = await createApiKey({
```

Comment:
Having this key created ad hoc is good but not enough. cast keys should be ephemeral and scoped bc they are handed over to the cast device. The cast device is inherently untreatable bc it is meant to be a common device sometimes in a public space.

Also the cast device should not have the view context of the current user.

A possible solution is to create an ephemeral "cast" shared album (instead of "cast" key) scoped with the photos in the cast session (added upon request) and deleted on session end with a possible expiry limitation.

Reply:
Yes, this is just the initial ad-hoc implementation. We were just discussing today in the discord how to get this better before merging the PR. Obviously this must be improved. One idea is to use jwt tokens with decent duration to send with the URL instead of the current API keys

Reply:
jwt is interesting but does the server already have support? 

creating a cast album may open the door to cast queues and even shared queues similar to youtube's casting.
also takes care of scope of visible scope. adding a key (riding on the password mechanism of shared album) and attaching the credentials to the cast device session could be a nice addition.